README.md
[![SnertSoft: We Serve Your Server](Img/logo-300x74.png)](http://software.snert.com/)

milter-length  « When Size Matters »
====================================

Copyright 2004, 2024 by Anthony Howe.  All rights reserved.


WARNING
-------

THIS IS MAIL FILTERING SOFTWARE AND WILL BLOCK MAIL THAT FAILS TO PASS A GIVEN SET OF TESTS.  SNERTSOFT AND THE AUTHOR DO NOT ACCEPT ANY RESPONSIBLITY FOR MAIL REJECTED OR POSSIBLE LOSS OF BUSINESSS THROUGH THE USE OF THIS SOFTWARE.  BY INSTALLING THIS SOFTWARE THE CLIENT UNDERSTANDS AND ACCEPTS THE RISKS INVOLVED.


Description
-----------


This is a [Sendmail](http://www.sendmail.org/) utility milter that imposes message size limits by IP address, domain name, or sender address on a message body length, excluding the message headers.  Sendmail's `MaxMessageSize` option only allows for a single global server wide message size limit, which is insufficient for some sites that would prefer finer granularity in the application of message size limits.  This is particularly useful for mail hosts that manage several domains and/or a large number of users, such as an ISP.

Note that if the Sendmail option `MaxMessageSize` is specified in the sendmail.cf configuration, then that will impose an upper limit to any `milter-length` message size limits. 


Usage
-----

        milter-length [options ...] [arguments ...]

Options can be expressed in four different ways.  Boolean options are expressed as `+option` or `-option` to turn the option on or off respectively.  Options that required a value are expressed as `option=value` or `option+=value` for appending to a value list.  Note that the `+option` and `-option` syntax are equivalent to `option=1` and `option=0` respectively.  Option names are case insensitive.

Some options, like `+help` or `-help`, are treated as immediate actions or commands.  Unknown options are ignored.  The first command-line argument is that which does not adhere to the above option syntax.  The special command-line argument `--` can be used to explicitly signal an end to the list of options.

The default options, as shown below, can be altered by specifying them on the command-line or within an option file, which simply contains command-line options one or more per line and/or on multiple lines.  Comments are allowed and are denoted by a line starting with a hash (#) character.  If the `file=` option is defined and not empty, then it is parsed first followed by the command-line options.

Note that there may be additional options that are listed in the option summary given by `+help` or `-help` that are not described here.


- - -
### access-db=/etc/mail/access.db

The type and location of the read-only access key-value map.  It provides a centralised means to black and white list hosts, domains, mail addresses, etc.  The following methods are supported:

        text!/path/map.txt                      R/O text file, memory hash
        /path/map.db                            Berkeley DB hash format
        db!/path/map.db                         Berkeley DB hash format
        db!btree!/path/map.db                   Berkeley DB btree format
        sql!/path/database                      An SQLite3 database
        socketmap!host:port                     Sendmail style socket-map
        socketmap!/path/local/socket            Sendmail style socket-map
        socketmap!123.45.67.89:port             Sendmail style socket-map
        socketmap![2001:0DB8::1234]:port        Sendmail style socket-map

If `:port` is omitted, the default is `7953`.

The `access-db` contains key-value pairs.  Lookups are performed from most to least specific, stopping on the first entry found.  Keys are case-insensitive.

An IPv4 lookup is repeated several times reducing the IP address by one octet from right to left until a match is found.

        tag:192.0.2.9
        tag:192.0.2
        tag:192.0
        tag:192

An IPv6 lookup is repeated several times reducing the IP address by one 16-bit word from right to left until a match is found.

        tag:2001:0DB8:0:0:0:0:1234:5678
        tag:2001:0DB8:0:0:0:0:1234
        tag:2001:0DB8:0:0:0:0
        tag:2001:0DB8:0:0:0
        tag:2001:0DB8:0:0
        tag:2001:0DB8:0:0
        tag:2001:0DB8:0
        tag:2001:0DB8
        tag:2001

A domain lookup is repeated several times reducing the domain by one label from left to right until a match is found.

        tag:[ipv6:2001:0DB8::1234:5678]
        tag:[192.0.2.9]
        tag:sub.domain.tld
        tag:domain.tld
        tag:tld
        tag:

An email lookup is similar to a domain lookup, the exact address is first tried, then the address's domain, and finally the local part of the address.

        tag:account@sub.domain.tld
        tag:sub.domain.tld
        tag:domain.tld
        tag:tld
        tag:account@
        tag:

If a key is found and is a milter specific tag (ie. `milter-length-Connect`, `milter-length-To`), then the value is processed as a pattern list and the result returned.  The Sendmail variants cannot have a pattern list.  A pattern list is a whitespace separated list of _pattern-action_ pairs followed by an optional default _action_.  The supported patterns are:

        [network/cidr]size            Classless Inter-Domain Routing
        !pattern!size                 Simple fast text matching.
        /regex/size                   POSIX Extended Regular Expressions

The CIDR will only ever match for IP address related lookups.

A `!pattern!` uses an astrisk (\*) for a wildcard, scanning over zero or more characters; a question-mark (?) matches any single character; a backslash followed by any character treats it as a literal (it loses any special meaning).

        !abc!           exact match for 'abc'
        !abc*!          match 'abc' at start of string
        !*abc!          match 'abc' at the end of string
        !abc*def!       match 'abc' at the start and match 'def' at the end, maybe with stuff in between.
        !*abc*def*!     find 'abc', then find 'def'

Below is a list of supported tags.  Other options may specify additional tags.

        milter-length-Connect:client-ip          size           § Can be a pattern list.
        milter-length-Connect:[client-ip]        size           § Can be a pattern list.
        milter-length-Connect:client-domain      size           § Can be a pattern list.
        milter-length-Connect:                   size           § Can be a pattern list.

        milter-length-Auth:auth_authen           size           § Can be a pattern list.
        milter-length-Auth:                      size           § Can be a pattern list.

        milter-length-From:sender-address        size           § Can be a pattern list.
        milter-length-From:sender-domain         size           § Can be a pattern list.
        milter-length-From:sender@               size           § Can be a pattern list.
        milter-length-From:                      size           § Can be a pattern list.

        milter-length-To:recipient-address       size           § Can be a pattern list.
        milter-length-To:recipient-domain        size           § Can be a pattern list.
        milter-length-To:recipient@              size           § Can be a pattern list.
        milter-length-To:                        size           § Can be a pattern list.

The _size_ is the maximum number of octets permitted in a message and is expressed as a number with an optional scale suffix: K (kilo), M (mega), or G (giga).  If no size is specified (or -1), then the message can be any length (ULONG_MAX).

When there are multiple message size limits assigned, then the limit applied, in order of precedence, will be: the maximum value of all relevant `milter-length-to:`, `milter-length-auth:`, `milter-length-from:`, `milter-length-connect:`, or an unlimited default.  Essentially each matching milter tag replaces the previous value.  When there are multiple recipients with one or more possible `milter-length-to:` entries, then maximum length limit found is applied. 

Below are only some examples of what is possible: 

        milter-length-Connect:80.94             [80.94.96.0/20]2M  500K
 
All the mail from netblock 80.94.96.0/20 (80.94.96.0 through to 80.94.111.255) is limited to 2 megabytes, while anything else in 80.94.0.0/16 is limited to 500 kilobytes.
 
        milter-length-Connect:192.0.2           10M
 
All the mail from the network in 192.0.2.0/24 is limited to 10 megabytes.
 
        milter-length-Connect:192.0.2           /^192\.0\.2\.8[0-9]/1M  10M

Mail sent from IP addresses 192.0.2.80 through to 192.0.2.89 is limited to 1 megabyte, while the rest of the network in 192.0.2.0/24 is limited to 10 megabytes.  This example is rather contrived to show some of what is possible with patterns and is the equivalent of ten simple `milter-length-connect:` entries.  When using patterns, weigh the need verses simplicity, readability, and administration.
 
        milter-length-From:example.com          /^[^+]+@/500K
 
Mail from within example.com, where the sender's address does not contain a plus-detail, is limited to 500 kilobytes, otherwise no limit is imposed.
 
        milter-length-To:example.net            !*+*@*!  /^[0-9].*/20K  !*smith*@*!1M 4M
 
Mail to addresses within `example.net` containing a plus-detail have no limit; those starting with a digit are limited to 20 kilobytes; mail to anyone with "smith" as part of their address is limited to 1 megabyte, while the rest of the domain is limited to 4 megabytes. 


- - -
### +daemon

Start as a background daemon or foreground application.


- - -
### file=/etc/mail/milter-length.cf

Read the option file before command line options.  This option is set by default.  To disable the use of an option file, simply say `file=''`.


- - -
### -help or +help

Write the option summary to standard output and exit.  The output is suitable for use as an option file.


- - -
### milter-socket=unix:/var/run/milter/milter-length.socket

A socket specifier used to communicate between Sendmail and `milter-length`.  Typically a unix named socket or a host:port.  This value must match the value specified for the `INPUT_MAIL_FILTER()` macro in the sendmail.mc file.  The accepted syntax is:

        {unix|local}:/path/to/file              A named pipe. (default)
        inet:port@{hostname|ip-address}         An IPV4 socket.
        inet6:port@{hostname|ip-address}        An IPV6 socket.


- - -
### milter-timeout=7210

The sendmail/milter I/O timeout in seconds.


- - -
### pid-file=/var/run/milter/milter-length.pid

The file path of where to save the process-id.


- - -
### -quit or +quit

Quit an already running instance of the milter and exit.  This is equivalent to:

        kill -QUIT `cat /var/run/milter/milter-length.pid`.


- - -
### -restart or +restart

Terminate an already running instance of the milter before starting.


- - -
### run-group=milter

The process runtime group name to be used when started by root.

- - -
### run-user=milter

The process runtime user name to be used when started by root.


- - -
### -smtp-auth-ok

Allow SMTP authenticated senders to send unscanned mail.  See also the `milter-length-auth:` tag (`access-db=`) for finer granularity of control.


- - -
### verbose=info

A comma separated list of how much detail to write to the mail log.  Those mark with `§` have meaning for this milter.

        §  all          All messages
        §  0            Log nothing.
        §  info         General info messages. (default)
        §  trace        Trace progress through the milter.
        §  parse        Details from parsing addresses or special strings.
           debug        Lots of debug messages.
           dialog       I/O from Communications dialog
           state        State transitions of message body scanner.
           dns          Trace & debug of DNS operations
           cache        Cache get/put/gc operations.
        §  database     Sendmail database lookups.
           socket-fd    socket open & close calls
           socket-all   All socket operations & I/O
        §  libmilter    libmilter engine diagnostics


- - -
### work-dir=/var/tmp

The working directory of the process.  Normally serves no purpose unless the kernel option that permits daemon process core dumps is set.


SMTP Responses
--------------

This is the list of possible SMTP responses:

* 550 5.7.1 max. message size \([0-9]+\) exceeded  
  The message body, not counting the headers, is too big.
    
* 553 5.1.0 imbalanced angle brackets in path  
  The path given for a `MAIL` or `RCPT` command is missing a closing angle bracket

* 553 5.1.0 address does not conform to RFC 2821 syntax  
  The address is missing the angle brackets, `<` and `>`, as required by the RFC grammar.

* 553 5.1.0 local-part too long  
  The stuff before the `@` is too long.

* 553 5.1.[37] invalid local part  
  The stuff before the `@` sign contains unacceptable characters.

* 553 5.1.0 domain name too long  
  The stuff after the `@` is too long.

* 553 5.1.7 address incomplete  
  Expecting a domain.tld after the `@` sign and found none.

* 553 5.1.[37] invalid domain name  
  The domain after the `@` sign contains unacceptable characters.


Build & Install
---------------

* Install `SQLite` from a package if desired.  Prior to [LibSnert's](https://github.com/SirWumpus/libsnert) availability on GitHub, the old `libsnert` tarballs included SQLite, but the GitHub [libsnert](https://github.com/SirWumpus/libsnert) repository does not, so it needs to be installed separately.   `milter-length` does not require it, but other milters that need a cache will.

* If you have never built a milter for Sendmail, then please make sure that you build and install `libmilter` (or install a pre-built package), which is _not_ built by default when you build Sendmail.  Please read the `libmilter` documentation.  Briefly, it should be something like this:

        cd (path to)/sendmail-8.13.6/libmilter
        sh Build -c install

* [Build LibSnert](https://github.com/SirWumpus/libsnert#configuration--build) first, do *not* disable `sqlite3` support; it should find the pre-installed version of SQLite if any.

* Building `milter-length` should be:

        cd com/snert/src
        git clone https://github.com/SirWumpus/milter-length.git
        cd milter-length
        ./configure --help
        ./configure
        make
        sudo make install

* An example `/usr/local/share/examples/milter-length/milter-length.mc` is supplied.  This file should be reviewed and the necessary elements inserted into your Sendmail `.mc` file and `sendmail.cf` rebuilt.  Please note the comments on the general milter flags.

* Once installed and configured, start `milter-length` and then restart Sendmail.  An example startup script is provided in `/usr/local/share/examples/milter-length/milter-length.sh`.


Notes
-----

* The minimum desired file ownership and permissions are as follows for a typical Linux system.  For FreeBSD, NetBSD, and OpenBSD the binary and cache locations may differ, but have the same permissions.  Process user `milter` is primary member of group `milter` and secondary member of group `smmsp`.  Note that the milter should be started as `root`, so that it can create a _.pid file_ and _.socket file_ in `/var/run`; after which it will switch process ownership to `milter:milter` before starting the accept socket thread.

        /etc/mail/                              root:smmsp      0750 drwxr-x---
        /etc/mail/access.db                     root:smmsp      0640 -rw-r-----
        /etc/mail/sendmail.cf                   root:smmsp      0640 -rw-r-----
        /etc/mail/milter-length.cf              root:root       0644 -rw-r--r--
        /var/run/milter/milter-length.pid       milter:milter   0644 -rw-r--r--
        /var/run/milter/milter-length.socket    milter:milter   0644 srw-r--r--
        /var/db/milter-length                   milter:milter   0644 -rw-r--r-- (*BSD)
        /var/cache/milter-length                milter:milter   0644 -rw-r--r-- (linux)
        /usr/local/libexec/milter-length        root:milter     0550 -r-xr-x---


CHANGES.TXT
CHANGES
-------

The major number changes for such things as code rewrites, changes in
syntax, and/or dramatic changes in functionality. The minor number
changes for corrections, enhancements, etc. made to the code. There
is no set schedule for releases.

--TODO--

--0.10--

   *	Requires LibSnert 1.75.8

   !	Configure and build changes to reflect changes in libsnert.

--0.9--

   !	Fixed possible off by one error for body chunks less than 20
   	bytes. Reported by Ben Lentz.

--0.8.22--

   !	Document correction. Removed old Usage section that no longer
   	applied. Reported by Ben Spencer.

--0.8--

   +	Requires LibSnert 1.63

  !!	filterMail() neglected to reset the maxBytesRcpt between messages.

   !	Corrected RFC 3463 Enhanced Mail System Status Codes from 5.7.1
   	to 5.3.4.

  !!	filterBody(): Fixed obsecure bug when the message content is
	less than 20 bytes long, adding a terminating NUL to the body
	chunk for the trace log output actually overwrote off the end
	of a dynamic buffer on some Linux systems causing the milter
	to crash. Reported by Martin Lathoud.

   +	access-db now supports socket-map, flat-file, and SQLite3
	lookups in addition to Berkeley DB. The SQLite3 behaviour
	can be disabled via LibSnert's ./configure script.

--0.7--

  !!	License 1.4 which provides additional details with regards to
  	possible license controls used, the possibility of "phone
  	home code" and its general contents, and related privacy
  	statement.

   +	Add check for filterOpen() got NULL socket address pointer passed to filterOpen()
   	which inidicates an unsupported address type. The connection
   	is simply accepted to bypass the milter.

--0.6--

   +	Requires LibSnert 1.61

  !!	Complete replacement of option specification and no backwards.
	compatibility. The newer format allows for better integration
	of previously undocumented libsnert options and for the
	inclusion of future enhancements.

   !	filterClose(): remove reporting of "null workspace error",
	which can occur when a client connection is rejected before or
	by filterOpen().

   !	Minor corrections for gcc -Wall warnings.

   !	Fixed smfAccessClient() argument bug.

--0.5.15--

  !!	Lets try this again...

   !	Fix rejection message which cited the wrong maxBytes variable
   	in the response. Reported by Doron Shmaryahu.


--0.5--

   !	Fix rejection message which cited the wrong maxBytes variable
   	in the response. Reported by Doron Shmaryahu.

   !	Allow for both upper and lower case unit specifiers, ie: 800k
   	and 800K should be the same. Reported by Doron Shmaryahu.

   !	Reverted back to libmilter's default (7210s) timeout. This
	should silence many of the reported sendmail to milter timeout
	errors and/or broken connections caused by the milter timing
	out sooner than sendmail does with the current SMTP client.
	The sendmail defaults for each SMTP command and receiving the
	message content is 1h each and so libmilter allows for 2h 10s.
	But in theory you could be DoS for between 5 and 104 hours per
	connection. If you want to change this then set in
	sendmail.mc:

		define(`confTO_DATABLOCK',      `30m')
		define(`confTO_COMMAND',        `5m')

--0.4--

   !	Fixed bug in selecting the max. message length to apply. The
   	choice is now based on the following precendence, from high
   	to low order:

  	  a) max. value of all relevant milter-length-to:,
  	  b) milter-length-auth:,
  	  c) milter-length-from:,
  	  d) milter-length-connect:,
  	  e) unlimited

   !	Replaced use of LIBSNERT_WITH_BERKELEY_DB by HAVE_DB_H due to
   	changes in LibSnert/1.59.

--0.3--

   +	Requires LibSnert 1.57

   +	Added support for default max. message length:

   		milter-length-connect:		RHS
   		milter-length-from:		RHS

   	Requested by Michael Long.

   +	Added support for

   		milter-length-auth:authid	RHS
   		milter-length-auth:		RHS

   	Requested by Michael Long.

   +	Added support for

   		milter-length-to:recipient	RHS
   		milter-length-to:		RHS

   	Requested by Sebastian Utz.

   +	Added host/IP to rejection message. Request by Michael Long.

--0.2--

   +	Requires LibSnert 1.56

   !	Changed install.sh.in to create /var/run/mitler/ directory
   	owned by milter:milter and change the default PID_FILE and
   	SOCKET_FILE paths. This provides better privelage/security
   	separation since the milters run as milter:milter.

--0.1--

   +	Requires LibSnert 1.53





milter-length.c
/*
 * milter-length.c
 *
 * Copyright 2005, 2012 by Anthony Howe. All rights reserved.
 *
 * The following should be added to the sendmail.mc file:
 *
 *	INPUT_MAIL_FILTER(
 *		`milter-length',
 *		`S=unix:/var/lib/milter-length/socket, T=S:30s;R:3m'
 *	)dnl
 */

/***********************************************************************
 *** Leave this header alone. Its generate from the configure script.
 ***********************************************************************/

#include "config.h"

/***********************************************************************
 *** You can change the stuff below if the configure script doesn't work.
 ***********************************************************************/

#ifndef RUN_AS_USER
#define RUN_AS_USER			"milter"
#endif

#ifndef RUN_AS_GROUP
#define RUN_AS_GROUP			"milter"
#endif

#ifndef MILTER_CF
#define MILTER_CF			"/etc/mail/" MILTER_NAME ".cf"
#endif

#ifndef PID_FILE
#define PID_FILE			"/var/run/milter/" MILTER_NAME ".pid"
#endif

#ifndef SOCKET_FILE
#define SOCKET_FILE			"/var/run/milter/" MILTER_NAME ".socket"
#endif

#ifndef WORK_DIR
#define WORK_DIR			"/var/tmp"
#endif

#ifndef SENDMAIL_CF
#define SENDMAIL_CF			"/etc/mail/sendmail.cf"
#endif

/***********************************************************************
 *** No configuration below this point.
 ***********************************************************************/

/* Re-assert this macro just in case. May cause a compiler warning. */
#define _REENTRANT	1

#include <com/snert/lib/version.h>

#include <sys/types.h>
#include <limits.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#include <com/snert/lib/berkeley_db.h>
#include <com/snert/lib/mail/limits.h>
#include <com/snert/lib/mail/smf.h>
#include <com/snert/lib/mail/smdb.h>
#include <com/snert/lib/util/Text.h>
#include <com/snert/lib/util/getopt.h>

#if LIBSNERT_MAJOR < 1 || LIBSNERT_MINOR < 75
# error "LibSnert 1.75.8 or better is required"
#endif

#ifndef HAVE_DB_H
# error "LibSnert must be built with Berkeley DB support"
#endif

#ifdef MILTER_BUILD_STRING
# define MILTER_STRING	MILTER_NAME"/"MILTER_VERSION"."MILTER_BUILD_STRING
#else
# define MILTER_STRING	MILTER_NAME"/"MILTER_VERSION
#endif

/***********************************************************************
 *** Constants
 ***********************************************************************/

#define	TAG_FORMAT		"%05d %s: "
#define	TAG_ARGS		data->work.cid, data->work.qid

#define X_SCANNED_BY		"X-Scanned-By"

/***********************************************************************
 *** Global Variables
 ***********************************************************************/

typedef struct {
	smfWork work;
	unsigned long tally;			/* per message */
	unsigned long maxBytes;			/* per message */
	unsigned long maxBytesRcpt;		/* per message */
	unsigned long maxBytesConnection;	/* per connection */
	char line[SMTP_TEXT_LINE_LENGTH+1];	/* general purpose */
	char client_name[SMTP_DOMAIN_LENGTH+1];	/* per connection */
	char client_addr[IPV6_TAG_LENGTH+IPV6_STRING_LENGTH];	/* per connection */
} *workspace;

static Option optIntro		= { "",	NULL, "\n# " MILTER_NAME "/" MILTER_VERSION "." MILTER_BUILD_STRING "\n#\n# " MILTER_COPYRIGHT "\n#\n" };

static Option *optTable[] = {
	&optIntro,
	NULL
};

/***********************************************************************
 *** Handlers
 ***********************************************************************/

static unsigned long
getMsgSizeLimit(workspace data, char *value)
{
	char *stop;
	unsigned long size;

	size = ULONG_MAX;

	if (value != NULL) {
		size = strtoul(value, &stop, 10);

		switch (*stop) {
		case 'G': case 'g': size *= 1024;
		case 'M': case 'm': size *= 1024;
		case 'K': case 'k': size *= 1024;
		}
	}

	smfLog(SMF_LOG_DEBUG, TAG_FORMAT "getMsgSizeLimit(%lx, %s) size=%lu", TAG_ARGS, (long) data, value, size);

	return size;
}

/*
 * Open and allocate per-connection resources.
 */
static sfsistat
filterOpen(SMFICTX *ctx, char *client_name, _SOCK_ADDR *raw_client_addr)
{
	workspace data;
	char *value = NULL;

	if (raw_client_addr == NULL) {
		smfLog(SMF_LOG_TRACE, "filterOpen() got NULL socket address, accepting connection");
		goto error0;
	}

	if (raw_client_addr->sa_family != AF_INET
#ifdef HAVE_STRUCT_SOCKADDR_IN6
	&& raw_client_addr->sa_family != AF_INET6
#endif
	) {
		smfLog(SMF_LOG_TRACE, "filterOpen() unsupported socket address type, accepting connection");
		goto error0;
	}

	if ((data = calloc(1, sizeof *data)) == NULL)
		goto error0;

	data->work.ctx = ctx;
	data->work.qid = smfNoQueue;
	data->work.cid = smfOpenProlog(ctx, client_name, raw_client_addr, data->client_addr, sizeof (data->client_addr));

	smfLog(SMF_LOG_TRACE, TAG_FORMAT "filterOpen(%lx, '%s', [%s])", TAG_ARGS, (long) ctx, client_name, data->client_addr);

	data->line[0] = '\0';
	data->work.replyLine[0] = '\0';
	data->maxBytesConnection = ULONG_MAX;

	if (smfi_setpriv(ctx, (void *) data) == MI_FAILURE) {
		syslog(LOG_ERR, "failed to save workspace");
		goto error1;
	}

	(void) TextCopy(data->client_name, sizeof (data->client_name), client_name);

	/* Lookup
	 *
	 *	milter-length-connect:a.b.c.d			RHS
	 *	milter-length-connect:a.b.c			RHS
	 *	milter-length-connect:a.b			RHS
	 *	milter-length-connect:a				RHS
	 *
	 *	milter-length-connect:ipv6:a:b:c:d:e:f:g:h	RHS
	 *	milter-length-connect:ipv6:a:b:c:d:e:f:g	RHS
	 *	milter-length-connect:ipv6:a:b:c:d:e:f		RHS
	 *	milter-length-connect:ipv6:a:b:c:d:e		RHS
	 *	milter-length-connect:ipv6:a:b:c:d		RHS
	 *	milter-length-connect:ipv6:a:b:c		RHS
	 *	milter-length-connect:ipv6:a:b			RHS
	 *	milter-length-connect:ipv6:a			RHS
	 *
	 *	milter-length-connect:[ip]			RHS
	 *	milter-length-connect:[ipv6:ip]			RHS
	 *	milter-length-connect:some.sub.domain.tld	RHS
	 *	milter-length-connect:sub.domain.tld		RHS
	 *	milter-length-connect:domain.tld		RHS
	 *	milter-length-connect:tld			RHS
	 *	milter-length-connect:				RHS
	 */
	if (smfAccessClient(&data->work, MILTER_NAME "-connect:", data->client_name, data->client_addr, NULL, &value) != SMDB_ACCESS_NOT_FOUND) {
		data->maxBytesConnection = getMsgSizeLimit(data, value);
		free(value);
	}

	return SMFIS_CONTINUE;
error1:
	free(data);
error0:
	return SMFIS_ACCEPT;
}

static sfsistat
filterMail(SMFICTX *ctx, char **args)
{
	workspace data;
	ParsePath *path;
	const char *error;
	char *value = NULL, *auth_authen;

	if ((data = (workspace) smfi_getpriv(ctx)) == NULL)
		return smfNullWorkspaceError("filterMail");

	if ((data->work.qid = smfi_getsymval(ctx, "i")) == NULL)
		data->work.qid = smfNoQueue;

	data->tally = 0;
	data->maxBytesRcpt = 0;
	data->maxBytes = data->maxBytesConnection;
	auth_authen = smfi_getsymval(ctx, smMacro_auth_authen);

	smfLog(SMF_LOG_TRACE, TAG_FORMAT "filterMail(%lx, %lx) MAIL='%s' auth='%s'", TAG_ARGS, (long) ctx, (long) args, args[0], auth_authen);

	if (args[0] == NULL)
		return SMFIS_CONTINUE;

	if (smfOptSmtpAuthOk.value && auth_authen != NULL) {
		smfLog(SMF_LOG_INFO, TAG_FORMAT "authenticated \"%s\", accept", TAG_ARGS, auth_authen);
		return SMFIS_ACCEPT;
	}

	if ((error = parsePath(args[0], smfFlags, 1, &path)) != NULL)
		return smfReply(&data->work, 553, NULL, error);

	/* Skip the null address. */
	if (*path->address.string == '\0')
		;

	/* Lookup
	 *
	 *	milter-length-auth:auth_authen			RHS
	 *	milter-length-auth:				RHS
	 */
	else if (smfAccessAuth(&data->work, MILTER_NAME "-auth:", auth_authen, path->address.string, NULL, &value) != SMDB_ACCESS_NOT_FOUND) {
		data->maxBytes = getMsgSizeLimit(data, value);
	}

	/* Lookup
	 *
	 *	milter-length-from:account@some.sub.domain.tld	RHS
	 *	milter-length-from:some.sub.domain.tld		RHS
	 *	milter-length-from:sub.domain.tld		RHS
	 *	milter-length-from:domain.tld			RHS
	 *	milter-length-from:tld				RHS
	 *	milter-length-from:account@			RHS
	 *	milter-length-from:				RHS
	 */
	else if (smfAccessEmail(&data->work, MILTER_NAME "-from:", path->address.string, NULL, &value) != SMDB_ACCESS_NOT_FOUND) {
		data->maxBytes = getMsgSizeLimit(data, value);
	}

	free(value);
	free(path);

	return SMFIS_CONTINUE;
}

static sfsistat
filterRcpt(SMFICTX *ctx, char **args)
{
	workspace data;
	ParsePath *path;
	const char *error;
	char *value = NULL;
	unsigned long messageSizeLimit;

	if ((data = (workspace) smfi_getpriv(ctx)) == NULL)
		return smfNullWorkspaceError("filterRcpt");

	smfLog(SMF_LOG_TRACE, TAG_FORMAT "filterRcpt(%lx, %lx) RCPT='%s'", TAG_ARGS, (long) ctx, (long) args, args[0]);

	if (args[0] == NULL)
		return SMFIS_CONTINUE;

	if ((error = parsePath(args[0], smfFlags, 0, &path)) != NULL)
		return smfReply(&data->work, 553, NULL, error);

	/* Lookup
	 *
	 *	milter-length-to:account@some.sub.domain.tld	RHS
	 *	milter-length-to:some.sub.domain.tld		RHS
	 *	milter-length-to:sub.domain.tld			RHS
	 *	milter-length-to:domain.tld			RHS
	 *	milter-length-to:tld				RHS
	 *	milter-length-to:account@			RHS
	 *	milter-length-to:				RHS
	 */
	if (smfAccessEmail(&data->work, MILTER_NAME "-to:", path->address.string, NULL, &value) != SMDB_ACCESS_NOT_FOUND) {
		messageSizeLimit = getMsgSizeLimit(data, value);
		if (data->maxBytesRcpt < messageSizeLimit)
			data->maxBytesRcpt = messageSizeLimit;
	}

	free(value);
	free(path);

	return SMFIS_CONTINUE;
}

static sfsistat
filterBody(SMFICTX *ctx, unsigned char *chunk, size_t size)
{
	workspace data;

	if ((data = (workspace) smfi_getpriv(ctx)) == NULL)
		return smfNullWorkspaceError("filterBody");

	data->tally += size;

	if (size == 0)
		chunk = "";
	else if (size < 20)
		chunk[--size] = '\0';

	smfLog(SMF_LOG_TRACE, TAG_FORMAT "filterBody(%lx, '%.20s...', %lu) tally=%lu", TAG_ARGS, (long) ctx, chunk, (unsigned long) size, data->tally);

	return SMFIS_CONTINUE;
}

static sfsistat
filterEndMessage(SMFICTX *ctx)
{
	workspace data;
	unsigned long maxBytes;

	if ((data = (workspace) smfi_getpriv(ctx)) == NULL)
		return smfNullWorkspaceError("filterEndMessage");

	smfLog(SMF_LOG_TRACE, TAG_FORMAT "filterEndMessage(%lx)", TAG_ARGS, (long) ctx);

	maxBytes = 0 < data->maxBytesRcpt ? data->maxBytesRcpt : data->maxBytes;

	if (maxBytes != ULONG_MAX && maxBytes < data->tally)
		return smfReply(&data->work, 550, "5.3.4", "%s [%s] (%lu bytes) exceeded max. message size of %lu bytes", data->client_name, data->client_addr, data->tally, maxBytes);

	smfLog(SMF_LOG_INFO, TAG_FORMAT "messages size %lu bytes", TAG_ARGS, data->tally);

	return SMFIS_CONTINUE;
}

/*
 * Close and release per-connection resources.
 */
static sfsistat
filterClose(SMFICTX *ctx)
{
	workspace data;
	unsigned short cid = 0;

	if ((data = (workspace) smfi_getpriv(ctx)) != NULL) {
		cid = smfCloseEpilog(&data->work);
		free(data);
	}

	smfLog(SMF_LOG_TRACE, TAG_FORMAT "filterClose(%lx)", cid, smfNoQueue, (long) ctx);

	return SMFIS_CONTINUE;
}

/***********************************************************************
 ***  Milter Definition Block
 ***********************************************************************/

static smfInfo milter = {
	MILTER_MAJOR,
	MILTER_MINOR,
	MILTER_BUILD,
	MILTER_NAME,
	MILTER_VERSION,
	MILTER_COPYRIGHT,
	NULL,
	NULL,
	NULL,
	NULL,
	NULL,
	NULL,
	SMF_STDIO_CLOSE,

	/* struct smfiDesc */
	{
		MILTER_NAME,		/* filter name */
		SMFI_VERSION,		/* version code -- do not change */
		0,			/* flags */
		filterOpen,		/* connection info filter */
		NULL,			/* SMTP HELO command filter */
		filterMail,		/* envelope sender filter */
		filterRcpt,		/* envelope recipient filter */
		NULL,			/* header filter */
		NULL,			/* end of header */
		filterBody,		/* body block filter */
		filterEndMessage,	/* end of message */
		NULL,			/* message aborted */
		filterClose		/* connection cleanup */
#if SMFI_VERSION > 2
		, NULL			/* Unknown/unimplemented commands */
#endif
#if SMFI_VERSION > 3
		, NULL			/* SMTP DATA command */
#endif
	}
};

/***********************************************************************
 *** Startup
 ***********************************************************************/

static void
atExitCleanUp()
{
	smdbClose(smdbAccess);
	smfAtExitCleanUp();
}

int
main(int argc, char **argv)
{
	int argi;

	/* Default is OFF. */
	smfOptSmtpAuthOk.initial = "-";

	smfOptFile.initial = MILTER_CF;
	smfOptPidFile.initial = PID_FILE;
	smfOptRunUser.initial = RUN_AS_USER;
	smfOptRunGroup.initial = RUN_AS_GROUP;
	smfOptWorkDir.initial = WORK_DIR;
	smfOptMilterSocket.initial = "unix:" SOCKET_FILE;

	/* Parse command line options looking for a file= option. */
	optionInit(optTable, smfOptTable, NULL);
	argi = optionArrayL(argc, argv, optTable, smfOptTable, NULL);

	/* Parse the option file followed by the command line options again. */
	if (smfOptFile.string != NULL && *smfOptFile.string != '\0') {
		/* Do NOT reset this option. */
		smfOptFile.initial = smfOptFile.string;
		smfOptFile.string = NULL;

		optionInit(optTable, smfOptTable, NULL);
		(void) optionFile(smfOptFile.string, optTable, smfOptTable, NULL);
		(void) optionArrayL(argc, argv, optTable, smfOptTable, NULL);
	}

	/* Show them the funny farm. */
	if (smfOptHelp.string != NULL) {
		optionUsageL(optTable, smfOptTable, NULL);
		exit(2);
	}

	if (smfOptQuit.string != NULL) {
		/* Use SIGQUIT signal in order to avoid delays
		 * caused by libmilter's handling of SIGTERM.
		 * smfi_stop() takes too long since it waits
		 * for connections to terminate, which could
		 * be a several minutes or longer.
		 */
		exit(pidKill(smfOptPidFile.string, SIGQUIT) != 0);
	}

	if (smfOptRestart.string != NULL) {
		(void) pidKill(smfOptPidFile.string, SIGQUIT);
		sleep(2);
	}

	(void) smfi_settimeout((int) smfOptMilterTimeout.value);
	(void) smfSetLogDetail(smfOptVerbose.string);

	openlog(MILTER_NAME, LOG_PID, LOG_MAIL);

	if (smfOptDaemon.value && smfStartBackgroundProcess())
		return 1;

	if (atexit(atExitCleanUp)) {
		syslog(LOG_ERR, "atexit() failed\n");
		return 1;
	}

	if (*smfOptAccessDb.string != '\0') {
		if (smfLogDetail & SMF_LOG_DATABASE)
			smdbSetDebugMask(SMDB_DEBUG_ALL);

		if ((smdbAccess = smdbOpen(smfOptAccessDb.string, 1)) == NULL) {
			syslog(LOG_ERR, "failed to open \"%s\"", smfOptAccessDb.string);
			return 1;
		}
	}

	return smfMainStart(&milter);
}
